processor/src/main/java/org/mapstruct/ap/internal/processor/MapperAnnotatedFormattingMessenger.java

```diff
@@ -0,0 +1,137 @@
+    }
+
+    @Override
+    public void printMessage(Element e, Message msg, Object... args) {
+        delegateMessager
+                        .printMessage(
+                            determineDelegationElement( e ),
+                            determineDelegationMessage( e, msg ),
+                            determineDelegationArguments( e, msg, args ) );
+    }
+
+    @Override
+    public void printMessage(Element e, AnnotationMirror a, Message msg, Object... args) {
+        delegateMessager
+                        .printMessage(
+                            determineDelegationElement( e ),
+                            a,
+                            determineDelegationMessage( e, msg ),
+                            determineDelegationArguments( e, msg, args ) );
+    }
+
+    @Override
+    public void printMessage(Element e, AnnotationMirror a, AnnotationValue v, Message msg, Object... args) {
+        delegateMessager
+                        .printMessage(
+                            determineDelegationElement( e ),
+                            a,
+                            v,
+                            determineDelegationMessage( e, msg ),
+                            determineDelegationArguments( e, msg, args ) );
+    }
+
+    @Override
+    public void note(int level, Message log, Object... args) {
+        delegateMessager.note( level, log, args );
+    }
+
+    @Override
+    public boolean isErroneous() {
+        return delegateMessager.isErroneous();
+    }
+
+    private Object[] determineDelegationArguments(Element e, Message msg, Object[] args) {
+        if ( methodInMapperClass( e ) ) {
+            return args;
+        }
+        String originalMessage = String.format( msg.getDescription(), args );
+        return new Object[] { originalMessage, constructMethod( e ), e.getEnclosingElement().getSimpleName() };
+    }
+
+    /**
+     * ExecutableElement.toString() has different values depending on the compiler. Constructing the method itself
+     * manually will ensure that the message is always traceable to it's source.
+     */
+    private String constructMethod(Element e) {
+        if ( e instanceof ExecutableElement ) {
+            ExecutableElement ee = (ExecutableElement) e;
+            StringBuilder method = new StringBuilder();
+            method.append( typeUtils.asElement( ee.getReturnType() ).getSimpleName() );
+            method.append( " " );
+            method.append( ee.getSimpleName() );
+            method.append( "(" );
+            boolean first = true;
+            for ( VariableElement element : ee.getParameters() ) {
+                if ( !first ) {
+                    method.append( ", " );
+                }
+                method.append( typeUtils.asElement( element.asType() ).getSimpleName() );
+                method.append( " " );
+                method.append( element.getSimpleName() );
+                first = false;
+            }
```

Comment:
We can change the loop with a stream and a collectors joining in theory. You could even add a method that would get the `VariableElement` and return its simple name type and then use a method reference.
e.g.

```suggestion
            method.append( ee.getParameters()
                    .stream()
                    .map( element -> typeUtils.asElement( element.asType() ).getSimpleName() )
                    .collect( Collectors.joining( ", " ) ) );
```

I think that this makes the code slightly clearer to read

Reply:
Well you did find the one bit of code that I was not that happy with ;)
The method is too long in my opinion, but then again we have a lot of methods in mapstruct that are longer. I don't mind changing it into the streaming API.

The next question here is what do we want as the method text, as your suggestion here strips the parameter names away.
I'm fine either way, but I believe it to be more clear to report with:
`<ReturnType> <MethodName>(<Param1Type> <Param1Name>, <Param2Type> <Param2Name>...).`
But leaving the parameter names behind reduces it to the actual signature (except that I add the return type as well ;) ) :
`<ReturnType> <MethodName>(<Param1Type>, <Param2Type>...)`

What do you think is better here, @filiphr ?

Reply:
I completely missed the parameter name. My suggestion is wrong then, you should include it. My main suggestion was for the style of the method, i.e. use streaming / StringJoiner instead of the loop, it makes it easier to read it.

I personally do not mind the length that much. We do try to extract smaller methods where it makes sense, but it doesn't work always 😀